_posts/2020-05-08-coupling-and-cohesion.markdown

```diff
@@ -0,0 +1,138 @@
+key principles.
+
+## Coupling
+
+Of the two concepts,
+[coupling](https://en.wikipedia.org/wiki/Coupling_%28computer_programming%29)
+is the one that most programmers are more likely to be familiar with.
+Coupling refers to the degree to which two components - classes or
+modules - interact with one another. It's a measure of how much they
+"know" about each other. When you're reminded to keep methods that
+aren't called by other classes private, you're hearing a half-hearted
+stab at coupling. Marking methods private is just one way to reduce
+the number of methods that any other part of the code can see or know
+about. But coupling can also occur through data - which is why using
+a global variable, for instance, is a code smell. Any class or module
+that references the global variable is now coupled to all of the
+others through that variable. Generally speaking, we want to reduce
+coupling to the bare minimum needed to implement a solution.
+Conveniently, there is a hierarchy of types of coupling. Personally,
+I've found that learning about this hierarchy, and at least keeping it
+in mind when designing software is worth the effort.
+
+
+## Cohesion
+
+The twin idea of Cohesion, for whatever reason, seems to be discussed
+less often, probably because it can be a bit more abstract at first.
+It's really just the inverse concept of coupling.
+[Cohesion](https://en.wikipedia.org/wiki/Cohesion_(computer_science))
+is the degree to which all of the methods and data structures in a
+class or module are related to one another and belong together.  A
+module or class with a high level of cohesion will have elements that
+all share a common purpose, while one with lower cohesion will be more
+of a loosly organized collection of odds and ends. Perhaps one reason
+Cohesion is not discussed as much is that it's easy to fall into the
+trap of thinking that a highly cohesive set of methods must be tightly
+coupled to one another, and therefore that the concept itself is just
+a restatement of Coupling. This is not necessarily the case, though.
+Ideally, a set of methods in a module or class would be *both* loosely
+coupled to one another, *and* highly cohesive. Achieving that in the
+real world can be challenging, but it's an ideal worth keeping in
+mind. Like Coupling, there is a [hierarchy of different
+levels](https://it.toolbox.com/blogs/craigborysowich/design-principles-cohesion-050307)
+of cohesion. Like coupling, having an at least passing familiarity
+with the different levels of Cohesion will change the way that you
+approach writing new classes and modules, or refactoring existing
+ones.
+
+## Some examples
+
+Like I said earlier, once you really start to understand coupling and
+cohesion, you'll start to see how most good design practices are
+really just ways of minimizing coupling and / or maximizing cohesion.
+For instance, most (if not all) of the design patterns listed as
```

Comment:
@rdormer Could you break down all the examples into paragraphs or bullet points? It will make it easier to read.